vm/linking.cpp

```diff
@@ -15,16 +15,277 @@
+      }
+    }
+
+    jr_code_load record = {};
+    record.p.id = JIT_CODE_LOAD;
+    record.p.total_size = sizeof(record) + strlen(symbol) + 1 + length;
+    record.p.timestamp = get_timestamp();
+    record.pid = pid_;
+    record.tid = pid_; // TODO: Use the thread id.
+    record.vma = (uint64_t)address;
+    record.code_addr = (uint64_t)address;
+    record.code_size = length;
+    record.code_index = (uint64_t)address; // We don't re-JIT, so this should be fine.
+
+    fwrite(&record, sizeof(record), 1, file_);
+    fwrite(symbol, 1, strlen(symbol) + 1, file_);
+    fwrite(address, 1, length, file_);
+
+    fflush(file_);
+  }
+
+ private:
+  inline uint64_t get_timestamp() {
+    timespec ts;
+
+    if (use_arch_timestamp_) {
+      return __rdtsc();
+    }
+
+    if (clock_gettime(CLOCK_MONOTONIC, &ts)) {
+      return 0;
+    }
+
+    return ((uint64_t)ts.tv_sec * 1000000000) + ts.tv_nsec;
+  }
+
+  uint16_t get_elf_machine() {
+    uint16_t machine = 0; /* EM_NONE */
+
+    FILE* exe = fopen("/proc/self/exe", "r");
+    if (!exe) {
+      return machine;
+    }
+
+    struct {
+      uint8_t ident[16];
+      uint16_t type;
+      uint16_t machine;
+    } __attribute__((packed)) elf_header;
+
+    if (fread(&elf_header, sizeof(elf_header), 1, exe) != -1) {
+      machine = elf_header.machine;
+    }
+
+    fclose(exe);
+    return machine;
+  }
+
+ private:
+  pid_t pid_;
+  bool self_delete_;
+  char path_[255];
+  void *mmap_;
+  FILE* file_;
+
+  // When being used with Intel PT profiling, we need to use the CPU clock as our time source.
+  bool use_arch_timestamp_;
+};
+#endif
+
 CodeChunk
-sp::LinkCode(Environment* env, Assembler& masm)
+sp::LinkCode(Environment* env, Assembler& masm, const char* name, const CodeDebugMap& mapping)
 {
   if (masm.outOfMemory())
     return CodeChunk();
 
-  CodeChunk code = env->AllocateCode(masm.length());
-  if (!code.address())
+  auto length = masm.length();
+  CodeChunk code = env->AllocateCode(length);
+
+  auto address = code.address();
+  if (!address)
     return code;
 
-  masm.emitToExecutableMemory(code.address());
+  masm.emitToExecutableMemory(address);
+
+  // Some other debug info consumers we might want to implement here:
+  //
+  // * Intel VTune
+  //   https://github.com/intel/ittapi
+  //   Would give us profiling coverage on non-Linux.
+  //   Very similar input (and use case) as perf, requires linking Intel code in though.
+  //
+  // * GDB JIT Interface
+  //   https://sourceware.org/gdb/current/onlinedocs/gdb/JIT-Interface.html
+  //   Lets GDB show JIT frames when debugging, with source info.
+  //   Requires generating full ELF + DWARF objects in memory.
+
+#if defined(KE_LINUX)
+  // These have to be shared among all environments, as the filenames are only distinguished by PID.
+  // Once we support multiple environments per process we'll need some internal locking.
+  static std::unique_ptr<PerfJitFile> perf_jit_file = {};
```

Comment:
It'd be better to find a globally visible place to put these, but for now this is fine. You shouldn't need the initializer though.

Reply:
I can't find a good spot elsewhere for these - any suggestions? I don't think we have any naturally global objects (Environments are the top-level instance), so they'd need to be global vars or static members of something (probably Environment?)